### ISSUE TITLE
JSC_TYPE_MISMATCH: restricted index type for array keys should ignore known array functions

### ISSUE THREAD

**Issue by mcurland** [google/closure-compiler#1675]
```js
var arr = [1,2,3];
var x = arr['join'](',');
```

will give an error
JSC_TYPE_MISMATCH: restricted index type
found   : string
required: number

on the join call. This is reasonable in cases where the string expression is not a standard array.prototype function. However, the JavaScript language spec is clear that arr.join and arr['join'] are to be treated as equivalent language constructs. This error also precludes expressions like: `arr[CONDITION ? 'push' : 'pop']`.

This error should not be triggered if the index evaluates to any constant string that is a standard function on array.prototype. This would be a reasonable compromise that would cover most cases (array prototype can be extended, so could be an issue).

Why am I hitting this? I use a personal js style convention where anything public (that I don't want to minify with the closure compiler) is ['quoted']. For me, this is easier than providing the compiler service lists of public names or other code markup and makes it clear in the both the defining and calling code which items are public API (uses quotes, not minified) and which are private (use dot notation). Obviously, for names that don't get minified anyway (anything in the standard DOM/JavaScript nameset) I can just use the dot notation, but it is against the normal style of the code.

In any case AFAIK, per the js language spec, `arr['join'](str)` is not doing an indexed lookup on the object and should not be an error.

**Comment by dimvar** [comment]
> Obviously, for names that don't get minified anyway (anything in the standard DOM/JavaScript nameset) I can just use the dot notation

This is what we recommend. Special casing for string literals to cover this narrow use case is kind of an overkill, and it won't cover more complex expressions anyway, like `CONDITION ? 'push' : 'pop'` that you mention.

**Comment by ray007** [comment]
It's also throwing an error when adding custom properties to an array.

**Comment by ChadKillingsworth** [comment]
Try turning off strong typing on the array:

```js
/** @type {Array<?>} */
var arr = [1,2,3];
var x = arr['join'](',');
```

**Comment by ray007** [comment]
@ChadKillingsworth I tried to turn off strong typing for array, but that does not work.
`/** @type {?} */` does help, but then it's not recognized as array anymore.
